docs: split up docs templates into different pages

JQuinnie opened the pull request:
## Description

This PR addresses #21321 , it splits up the sub-secitons of the [Docs Templates](https://www.gatsbyjs.org/contributing/docs-templates/) into different pages for each type of template.

**Before**
![image](https://user-images.githubusercontent.com/29465981/74805670-06789700-52a1-11ea-9786-547fde5dd5ec.png)

**After**
![image](https://user-images.githubusercontent.com/29465981/74805691-1ee8b180-52a1-11ea-81e0-23cc6b731163.png)

#### Tasks
- [x] split and pull out docs templates sub sections into 5 different `md` files
- [x] update sidebar menu, `contributing-links.yaml` to include the new pages
- [x] update internal links to point to new pages

### Notes and Considerations

When developing, the newly created pages don't seem to load in the `PrevAndNext` component correctly... upon further investigation, it seems that `get-prev-and-next` does not pull in the updated [`contributing-links.yaml`](https://github.com/gatsbyjs/gatsby/blob/2c7212ea1df9a5e9563084783a0cab70c0f69f0e/www/src/utils/get-prev-and-next.js#L9-L11) file during `gatsby develop`, restarting does not seem to help either... so I am not sure if it is just an issue in development. If it is something that I need to address during this PR, please do let me know.

**Expect Result**
![image](https://user-images.githubusercontent.com/29465981/74806535-6f610e80-52a3-11ea-9096-04d2d88b35d0.png)

**Actual Result**
![image](https://user-images.githubusercontent.com/29465981/74806589-8f90cd80-52a3-11ea-9312-7e1c5da59db0.png)

## Related Issues

Addresses PR #21321

laurieontech commented:
Checked this locally and it looks good. Agree on the prev/next links. I'll tag in @AishaBlake and @tesseralis since they'll likely have more context.

tesseralis commented:
The prev/next links shouldn't be an issue. If you restarted and it didn't work, it's probably because it saved the prev/next page information in the cache. If you `rm -rf www/.cache` and rerun, it should work.

JQuinnie commented:
> The prev/next links shouldn't be an issue. If you restarted and it didn't work, it's probably because it saved the prev/next page information in the cache. If you `rm -rf www/.cache` and rerun, it should work.

Ah! That makes sense, thanks so much! 👍

JQuinnie commented on www/src/data/sidebars/contributing-links.yaml:
```diff
@@ -59,6 +59,16 @@
+            - title: How to Write a Plugin README
+              link: /contributing/how-to-write-a-plugin-readme/
+            - title: How to Write a Starter README
+              link: /contributing/how-to-write-a-starter-readme/
```
Sounds good, will update!

JQuinnie commented on docs/contributing/how-to-write-a-plugin-readme.md:
```diff
@@ -0,0 +1,63 @@
+---
+title: How to Write a Plugin README
```
Hi @tesseralis ,

Hope this is what you were looking for #21599 and #21600

tesseralis replied:
Perfect! Now just link it here:

```
title: How to Write a Plugin README
issue: #21600 
```

JQuinnie replied:
@tesseralis gotcha, just made these updates.

JQuinnie commented:
@tesseralis I have addressed all your comments, please have another look when you get the chance;

![image](https://user-images.githubusercontent.com/29465981/74889517-899bfa80-533e-11ea-8fca-fbdc437d75ee.png)

:)

marcysutton commented:
About the slug: we typically make them match page titles as it follows a predictable convention and helps with SEO. CC @hashimwarren.

JQuinnie commented:
Hi All,

Noticed that there's a CI build fail on multiple PRs for some time, not sure if its something that I need to do anything about? If there is a next step in this PR process please do let me know.

Thanks!

laurieontech commented:
Looks like there are a few more things needed before this goes in. There needs to be a redirect for `https://www.gatsbyjs.org/contributing/docs-templates/ ` since this PR removes it.

Additionally, all of the links within the larger body of docs need to be updated so they don't throw 404s.

JQuinnie commented:
> Looks like there are a few more things needed before this goes in. There needs to be a redirect for `https://www.gatsbyjs.org/contributing/docs-templates/ ` since this PR removes it.
> 
> Additionally, all of the links within the larger body of docs need to be updated so they don't throw 404s.

Hi @laurieontech ,

Thanks for the review, I will go ahead and check the links and make appropriate updates, please give me a few days as I am travelling currently. As always, appreciate your guidance and help!

JQuinnie commented:
> Looks like there are a few more things needed before this goes in. There needs to be a redirect for `https://www.gatsbyjs.org/contributing/docs-templates/ ` since this PR removes it.
> 
> Additionally, all of the links within the larger body of docs need to be updated so they don't throw 404s.

Hi @laurieontech , I have gone and review all the `docs-templates` links and fixed what I found, none threw a 404. If there are specific ones you have encountered, please let me know.

laurieontech commented:
Thanks. I can review that, but we still need the redirect for the page being removed.

JQuinnie commented:
Good morning @laurieontech ,

I took the `Docs Templates` page and split some of its sections into sub-pages underneath it, I don't believe I have removed any pages that would need a redirect.

`.../contributing/docs-templates/` loads up just fine when I spin up the service locally, I didn't remove this page itself.

Sorry if I am misunderstanding your guidance, could you be more specific on which page will need a redirect? Thanks!

laurieontech commented:
Ok. It looks like the way this was done is that the templates page still exists as a top-level explanation of the pages that follow. This might need more thought since that doesn't follow other conventions. Tagging in @marcysutton.

laurieontech commented:
I went through and looked for links we'll want to change. While they don't 404, they don't all point to the correct place anymore. For example `/contributing/docs-templates/#plugin-readme-template` in the "Submit to Plugin Library" doc sends you to the docs-template page rather than the plugin README template page. There are other links like that where you'd previously navigate to a header but now need to be sent to a different page. So instead of a redirect rule, this PR should include the updated links in those documents.

marcysutton commented on docs/contributing/docs-templates.md:
````diff
@@ -7,487 +7,13 @@ Docs reference guides, recipes, and tutorials teach Gatsby concepts to users wit
-
-Any tips on how to deploy this starter? What CMS to use? Other fun ways to build on top of the starter? Say those here!
-```
+Follow the how-to templates (models) listed here when contributing to Gatsby docs to ensure that the docs accomplish their purpose. If you have a good reason to deviate from the following template structures, mention those reasons in the PR so others can give proper feedback.
````
```suggestion
Follow the how-to templates (models) listed here when contributing to Gatsby docs to ensure that the docs accomplish their purpose. If you have a good reason to deviate from the following template structures, mention those reasons in an issue so others can give proper feedback on your idea.
```

We need to bias toward issues before PRs for discussion with @gatsbyjs/learning, especially if someone wants to deviate from expected formats.

With subpages under this overview page in `contributing-links`, you could also ideally use the `<Guidelist />` component to automatically list them: https://www.gatsbyjs.org/contributing/docs-and-blog-components/#guide-list

JQuinnie replied:
@marcysutton thank you so much for taking the time to go over these documents and providing feedback on each change and its necessity. I do not have the level and depth of each particular template/doc so this was a great learning experience for me. 😃

marcysutton commented on docs/contributing/gatsby-style-guide.md:
```diff
@@ -25,7 +25,7 @@ Before you begin writing, make sure to read the rest of this style guide.
 examples:
 
-- [Reference guides](/docs/add-404-page/)
+- [Reference guides](/docs/guides/)
```
This points to a specific example, not a section. I'd leave it as-is.

JQuinnie replied:
👍

marcysutton commented on docs/contributing/how-to-write-a-reference-guide.md:
````diff
@@ -0,0 +1,175 @@
+- Links to starters
+```
+
+---
````
I think this horizontal line can go away.
```suggestion
```

marcysutton commented on docs/contributing/how-to-write-a-plugin-readme.md:
````diff
@@ -0,0 +1,64 @@
+If you have unanswered questions, would like help with enhancing or debugging the plugin, it is nice to include instructions for people who want to contribute to your plugin.
+```
+
+---
````
```suggestion
```

marcysutton commented on docs/contributing/how-to-write-a-tutorial.md:
````diff
@@ -0,0 +1,100 @@
+- Links to starters
+```
+
+---
````
```suggestion
```

marcysutton commented on docs/contributing/submit-to-plugin-library.md:
```diff
@@ -9,7 +9,7 @@ In order to add your plugin to the [Plugin Library](/plugins/), you need to:
 2.  Include the [required files](/docs/files-gatsby-looks-for-in-a-plugin/) in your plugin code.
 3.  **Include a `keywords` field** in your plugin's `package.json`, containing `gatsby` and `gatsby-plugin`. If your plugin is a theme, please also include `gatsby-theme`.
-4.  Document your plugin with a README, there is a [plugin template](/contributing/docs-templates/#plugin-readme-template) available to be used as reference.
+4.  Document your plugin with a README, there is a [plugin template](/contributing/how-to-write-a-plugin-readme/#near-perfect-example-of-a-plugin-readme) available to be used as reference.
```
With a little rewording, the whole plugin README contributing doc should suffice here.

```suggestion
4.  Document your plugin with a README, using the contributing [plugin template](/contributing/how-to-write-a-plugin-readme/) for reference.
```

marcysutton commented on docs/contributing/how-to-write-a-reference-guide.md:
```diff
@@ -0,0 +1,175 @@
+Reference guide articles cover discrete topics as documentation with links to other resources. A reference guide explains a Gatsby concept or technique without the step-by-step context provided by a tutorial or recipe.
+
+Reference guide sections provide canonical information on how and why to build things with Gatsby for a variety of scenarios. Reference material should list text headings for each section of a guide, rather than "steps 1-5" like a tutorial.
+
```
```suggestion

## What are reference guide overview pages?

**Reference guide overview pages** act as parent pages for groups of reference guide articles, offering an original summary of what can be found in the section. They almost always include a `<Guidelist />` component to automatically list the subpage links in that section of the sidebar. See the [Docs and Blog Components](/contributing/docs-and-blog-components/#guide-list) page to learn more about the Guidelist component.
```

Since this doc now covers both things, we should mention both formats up front.

marcysutton commented on docs/contributing/translation/maintainers.md:
```diff
@@ -34,7 +34,7 @@ The repo creation script will create a progress issue listing the list of [core
 The Gatsby learning team is in charge of determining priorities for which docs should be translated. Refer to the [i18n page spreadsheet](https://docs.google.com/spreadsheets/d/1u2amGnqFLKxJuL5h9UrDblUueFgg0EBt7xbau4n8iTM/edit) to get the most up-to-date priority list, which includes frequently-visited pages in the Gatsby docs, tutorial, recipes, and other important pages.
 
-[Reference guide overview pages](/contributing/docs-templates/#reference-guide-overview) are also worth translating to establish a fully translated path to a frequently visited [reference guide](/contributing/docs-templates/#reference-guides), though they are listed at a lower priority.
+[Reference guide overview pages](/contributing/docs-templates/how-to-write-a-reference-guide/#reference-guide-template) are also worth translating to establish a fully translated path to a frequently visited [reference guide](/contributing/how-to-write-a-reference-guide/#what-are-reference-guide-articles), though they are listed at a lower priority.
```
Reference guide overviews and reference guides themselves serve different purposes, so we'll want to link to the correct sections.

```suggestion
[Reference guide overview pages](/contributing/docs-templates/how-to-write-a-reference-guide/#what-are-reference-guide-overview-pages) are also worth translating to establish a fully translated path to a frequently visited [reference guide](/contributing/how-to-write-a-reference-guide/#what-are-reference-guide-articles), though overview pages are listed at a lower priority.
```

marcysutton commented on www/src/data/sidebars/contributing-links.yaml:
```diff
@@ -63,6 +63,22 @@
               link: /contributing/docs-templates/
+              items:
+                - title: How to Write a Reference Guide
+                  breadcrumbTitle: Reference Guide
```
```suggestion
                  breadcrumbTitle: Reference Guides
```
I'd recommend making this plural since it includes overview pages, and we can make pluralization consistent throughout.

marcysutton commented on www/src/data/sidebars/contributing-links.yaml:
```diff
@@ -63,6 +63,22 @@
+                  breadcrumbTitle: Reference Guide
+                  link: /contributing/how-to-write-a-reference-guide/
+                - title: How to Write a Recipe
+                  breadcrumbTitle: Recipe
```
```suggestion
                  breadcrumbTitle: Recipes
```

marcysutton commented on www/src/data/sidebars/contributing-links.yaml:
```diff
@@ -63,6 +63,22 @@
+                  breadcrumbTitle: Recipe
+                  link: /contributing/how-to-write-a-recipe/
+                - title: How to Write a Tutorial
+                  breadcrumbTitle: Tutorial
```
```suggestion
                  breadcrumbTitle: Tutorials
```

marcysutton commented on www/src/data/sidebars/contributing-links.yaml:
```diff
@@ -63,6 +63,22 @@
+                  breadcrumbTitle: Tutorial
+                  link: /contributing/how-to-write-a-tutorial/
+                - title: How to Write a Plugin README
+                  breadcrumbTitle: Plugin README
```
```suggestion
                  breadcrumbTitle: Plugin READMEs
```

marcysutton commented on www/src/data/sidebars/contributing-links.yaml:
```diff
@@ -63,6 +63,22 @@
+                  breadcrumbTitle: Plugin README
+                  link: /contributing/how-to-write-a-plugin-readme/
+                - title: How to Write a Starter README
+                  breadcrumbTitle: Starter README
```
```suggestion
                  breadcrumbTitle: Starter READMEs
```

JQuinnie commented:
@marcysutton thank you so much for the review, I have made the requested changes to my PR. I really appreciate the help!
